freedb/protocols.md

```diff
@@ -1,8 +1,311 @@
+
+Suppose the client schema looks like this (written in `PyFreeDB` model):
+```
+class Person(models.Model):
+    name = models.StringField() # 1st column
+    age = models.IntegerField() # 2nd column
+```
+
+and we have data below:
+```
+[
+    Person(name="A", age=1), # 1st row
+    Person(name="B", age=2), # 2nd row
+    Person(name="C", age=3), # 3rd row
+    Person(name="D", age=4)  # 4th row
+]
+```
+
+Then the actual data that we store in Google Sheet would look like this:
+| _rid | name | age |
+| ---- | ---- | --- |
+| 1    | A    | 1   |
+| 2    | B    | 2   |
+| 3    | C    | 3   |
+| 4    | D    | 4   |
+
+#### About `_rid` Metadata
+
+`_rid` column will be used to tell the index of a row. This metadata will make our Google Sheet Formula simpler
+so that we can utilize [Google Visualization API][GVizAPI] for our needs.
+
+⚠️ Even though `_rid` can be used as the row identifier (because `_rid` is unique between each row), we recommend the
+client to implement their own row identifier mechanism such as [UUID](https://en.wikipedia.org/wiki/Universally_unique_identifier)
+or [Snowflake](https://en.wikipedia.org/wiki/Snowflake_ID) because we will not guarantee that `_rid` will not be
+reused out-of-the-box (to ensure the `_rid` is not reused, you need to implement soft-delete mechanism by your own).
+
+Consider this scenario:
+1. Client inserts Row A to the store, which will have `_rid=2`.
+2. Client deletes Row A from the store.
+3. Client Insert Row B to the store, which will reuse the same `_rid` as A i.e. B's `_rid` will be equal to `2`.
+
+We did not expose this field by default because if the client didn't aware of this behavior it can cause various bugs
+because of the invalid/dangling IDs.
+
+### Operations
+
+#### Insert
+
+Inserts new rows into the database.  The operation method should expect these parameters:
+| Parameter Name | Required | Remarks                                                   |
+| -------------- | -------- | --------------------------------------------------------- |
+| `rows`         | Yes      | List of row objects that will be inserted into the store. |
+
+
+Call [spreadsheets.values.append][AppendAPI]
+API with the following parameters:
+
+| Parameter Name              | Value                      |
+| --------------------------- | -------------------------- |
+| `spreadsheetId`             | `<current_spreadsheet_id>` |
+| `range`                     | `<data_range>`             |
+| `responseValueRenderOption` | `FORMATTED_VALUE`          |
+| `valueInputOption`          | `client_ENTERED`           |
+| `includeValuesInResponse`   | `true`                     |
+| `insertDataOption`          | `OVERWRITE`                |
+
+Replace the `<current_spreadsheet_id>` with the Google Spreadsheet ID that we're going to operate on and `<data_range>`
+with the range where the data lives in [A1 notation format][A1Notation]. It's recommended to select the entire column
```

Comment:
Recommended or must use? I feel it's a must?

Reply:
The recommendation is to select all columns, but specifying just `A1` should also worked. Do you think we should force reader to use the same formula?

Reply:
I think we should keep it uniform as much as possible (to make it less complicated to maintain and to reason also). I am just imagining if we have 5 libraries and each is using different ways (with the same result). It will be hard for us to keep track which is which.

I think we should say "very strongly recommended". But maybe we can have a warning section saying why it is not a must, but a strong recommendation.

Reply:
Updated, added warning on the top of the page.